### ISSUE TITLE
[Feature Request] Run `scripts` for all configured workspaces concurrently

### ISSUE THREAD

**Issue by cinderblock** [kimmobrunfeldt/concurrently#284]
Since Npn has added workspaces (and Yarn has them too), it would be nice to have a "concurrently" version of those.

For instance, with `npm run --workspaces test`, the `test` `package.json#scripts` will be run for all workspaces sequentially.

A "concurrently" version of the above might be: `concurrently --workspaces npm:test`.

**Comment by daniloraisi** [comment]
As I understand, this is more complex than creating a way to run commands from workspaces.

When you execute `npm run --workspaces test` or `yarn workspace test`, the command will be handled by Yarn and NPM to use the workspace context but referencing other things from the parent (such as node_modules).

I use concurrently with Yarn Workspaces.

Suppose you have a project with two packages (client and server) and have configured Yarn Workspaces (I'm considering you are using Yarn Berry).

I add Concurrently as a Dev Dependency for my root package.json. `yarn add -D concurrently`

Then, create scripts to execute server and client.

```json
{
  "scripts": {
    "dev:client": "yarn workspace client dev",
    "dev:server": "yarn workspace server dev",
    "dev": "concurrently \"yarn:dev:*\""
  }
}
```

**Comment by cinderblock** [comment]
Yes, that works. But the point of this feature request is basically that the list of `dev:*` scripts should not be necessary.

I'd (personally) rather `concurrently` automatically try each workspace's `dev` script, silently failing if one is missing (possibly). It seems to me this behaviour could be unlocked with a new cli option and not mess with any existing setups.

Throwing a wrench in here, Yarn and Npm handle workspaces completely differently...

**Comment by cinderblock** [comment]
Personally, I think using the `yarn:` or `npm:` prefix is confusing and overloading of namespaces.

Ideally, one could ignore that Yarn or Npm are used under the hood. `--workspaces` might be able to detect which one is currently in use, with an optional modifier like `--workspaces=yarn` or `--workspaces=npm` if you want to specify one or the other.

That being said, I don't think what you're proposing would work. Running `npm run test --workspaces` would run the `test` script on each workspace sequentially - exactly the opposite of what I'm trying to achieve. Concurrently basically needs to get the list of workspaces and run each `npm run test --workspace=foobar` for every one.

**Comment by daniloraisi** [comment]
The use of `npm:` or `yarn:` keeps the standard used by concurrently. The user specifies which package manager he is using.

With Yarn, we have a `--parallel` flag. As npm Workspaces is very recent, there's no option like that (yet). I don't think it is reasonable to create such a structure to validate if we are using Workspaces to run in parallel. In my opinion, It might be an unnecessary overhead.

Also, npm v7+ is available in Node v15+. To use it with Node LTS, we must install it manually. Some tools support Node LTS only with npm ^6.0.0, and migrating to npm v7 can break some things.

That said, I'll start a draft for this task and, as the development evolves, we can discuss the better way to solve things.

**Comment by adambubenicek** [comment]
For anyone looking for a quick workaround, this seems to be working fine for me

`.bin/concurrently`: (with `chmod +x`)
```javascript
#!/usr/bin/env node

const concurrently = require("concurrently")
const package = require("../package.json")
const script = process.argv[2]

concurrently(package.workspaces.map((workspace) => ({
    command: `npm run ${script} -w ${workspace} --if-present`,
    name: `${workspace}`
})))
```

and in `package.json`:
```json
{
  "scripts": {
    "watch": ".bin/concurrently watch"
  },
  "devDependencies": {
    "concurrently": "^7.0.0"
  }
}
```

**Comment by cinderblock** [comment]
@adambubenicek Nice. Better to use `node` binary directly so that you don't need `chmod +x` so that it works on Windows.

**Comment by gustavohenke** [comment]
This would be a cool addition, however, I don't think it can be implemented for all package managers:

- [npm](https://docs.npmjs.com/cli/v7/using-npm/workspaces) is pretty simple to work with
- [yarn](https://yarnpkg.com/features/workspaces) uses JSON, but accepts globs
- [pnpm](https://pnpm.io/pnpm-workspace_yaml) uses YAML and accepts globs
- [deno](https://docs.deno.com/runtime/fundamentals/workspaces/#multiple-package-entries) seems to be similar to npm? there's no evidence in the docs that it supports globs.

**Comment by plutonium-239** [comment]
this is quite important for monorepo tooling, any updates?

**Comment by gustavohenke** [comment]
See above.
